File: backend/internal/roles/roles.go
package roles

const (
	Reader      = "READER"
	Author      = "AUTHOR"
	Admin       = "ADMIN"
	SuperAdmin  = "SUPER_ADMIN"
)

var rank = map[string]int{
	Reader:     1,
	Author:     2,
	Admin:      3,
	SuperAdmin: 4,
}

func HasAtLeast(userRole, required string) bool {
	return rank[userRole] >= rank[required]
}

File: frontend/lib/roles.ts
const rank: Record<string, number> = {
  READER: 1,
  AUTHOR: 2,
  ADMIN: 3,
  SUPER_ADMIN: 4,
};

export const roles = {
  Reader: "READER",
  Author: "AUTHOR",
  Admin: "ADMIN",
  SuperAdmin: "SUPER_ADMIN",
  hasAtLeast(userRole: string, required: string) {
    return (rank[userRole] ?? 0) >= (rank[required] ?? 0);
  },
};
